joint_state_publisher/joint_state_publisher/joint_state_publisher.py

```diff
@@ -139,38 +139,116 @@ def init_urdf(self, robot):
                     joint['continuous'] = True
                 self.free_joints[name] = joint
 
-    def __init__(self):
-        description = get_param('robot_description')
-
-        self.free_joints = {}
-        self.joint_list = [] # for maintaining the original order of the joints
-        self.dependent_joints = get_param("dependent_joints", {})
-        self.use_mimic = get_param('use_mimic_tags', True)
-        self.use_small = get_param('use_smallest_joint_limits', True)
-
-        self.zeros = get_param("zeros")
-
-        self.pub_def_positions = get_param("publish_default_positions", True)
-        self.pub_def_vels = get_param("publish_default_velocities", False)
-        self.pub_def_efforts = get_param("publish_default_efforts", False)
-
+    def configure_robot(self, description):
+        self.node.get_logger().info('Got description, configuring robot')
         robot = xml.dom.minidom.parseString(description)
         if robot.getElementsByTagName('COLLADA'):
             self.init_collada(robot)
         else:
             self.init_urdf(robot)
 
-        source_list = get_param("source_list", [])
+    def parse_dependent_joints(self):
+        dj = {}
+        dependent_joints = self.node.get_parameters_by_prefix('dependent_joints')
+        # get_parameters_by_prefix returned a dictionary of keynames like
+        # 'head.parent', 'head.offset', etc. that map to rclpy Parameter values.
+        # The rest of the code assumes that the dependent_joints dictionary is
+        # a map of name -> dict['parent': parent, factor: factor, offset: offset],
+        # where both factor and offset are optional.  Thus we parse the values we
+        # got into that structure.
+        for name,param in dependent_joints.items():
+            # First split on the dots; there should be one and exactly one dot
+            split = name.split('.')
+            if len(split) != 2:
+                raise Exception("Invalid dependent_joint name '%s'" % (name))
+            newkey = split[0]
+            newvalue = split[1]
+            if newvalue not in ['parent', 'factor', 'offset']:
+                raise Exception("Invalid dependent_joint name '%s'" % (newvalue))
+            if newkey in dj:
+                dj[newkey].update({newvalue: param.value})
+            else:
+                dj.update({newkey: {newvalue: param.value}})
```

Comment:
I'm more used to seeing `dj[newkey] = {newvalue: param.value}` for the initial value

Reply:
Fixed in 2917803.